rails_programming/rails_sprinkles/css_bundling.md

```diff
@@ -0,0 +1,74 @@
+
+4. [PostCSS](https://postcss.org/)
+A tool for transforming CSS by using Javascript. It allows you to do things like use future CSS features not supported in all browsers or use one of the many postcss plugins to enhance your CSS.
+
+5. [Tailwind CSS](https://tailwindcss.com/)
+A utility-first CSS framework. It works in a similar way to bootstrap in that you add CSS to your application by adding classes to your HTML. Where it differs is Tailwind is not a way to avoid understanding CSS but instead allows you to build up your design by using repeatable classes that do only one thing. It's moving the CSS to the markup instead of under a class name in your CSS files. You'll either love it or hate it.
+
+### Installation
+
+For a new Rails 7 app installation is as simple as a command line argument when creating the app
+
+~~~bash
+rails new myapp --css [tailwind|bootstrap|bulma|postcss|sass]
+~~~
+
+So if you wanted to use bootstrap you'd run
+
+~~~bash
+rails new myapp --css bootstrap
+~~~
+
+For an existing application you just need to add `css-bundling-rails` to your Gemfile, run `bundle install` in the command line and then use the generator provided by the gem to install the tooling you need `./bin/rails css:install:[tailwind|bootstrap|bulma|postcss|sass]`, so again using bootstrap as an example you'd write `./bin/rails css:install:bootstrap`.
+
+That's all there is to it.
+
+### Build script
+
+When you use cssbundling it does a couple of things for you. Firstly, no matter which option you choose, it installs the necessary tools files necessary to get it running and makes sure the necessary import statements are added to your application. It then gives you instructions on the necessary build script you need to use.
+
+The build script is used to take the css you've written and process it through the tooling you've selected. So if you choose tailwind it needs to take all those utility classes and make sure only those that you use are included to keep the tailwind file as small as possible. The build script is placed in your package.json file which is why you do need nodejs installed to use this gem. In production the build script is used to ensure all your assets are built correctly but in development you need to rebuild everytime there is a change. Because of this the gem ships with a command used to watch your files and rebuild when necessary.
+
+If you use either jsbundling or cssbundling you should always start by running `./bin/dev` in your terminal. This is a task provided by Rails to run not only your Rails server but also the js and css watch commands to ensure your assets are rebuilt if they need to be. Under the hood that command uses a gem called [foreman](https://github.com/ddollar/foreman) which uses a Procfile to run lots of different processes in the same terminal so it can run the Rails server as well as watching over your assets for changes.
```

Comment:
Just a bit of suggested rewording, theres a lot to unpack with what the task does and how it works.

```suggestion
If you use the jsbundling or cssbundling gems, you should always start your app by running `./bin/dev` in your terminal. This task will run the Rails server, and the js and CSS watch commands to ensure your assets rebuild when needed. 

The  `./bin/dev` task uses a gem called [foreman](https://github.com/ddollar/foreman), which uses a Procfile to run lots of different processes in the same terminal, so that it can run the Rails server as well as watch over your assets for changes.
```